picard/ui/options/acousticbrainz.py

```diff
@@ -0,0 +1,146 @@
+
+
+class AcousticBrainzOptionsPage(OptionsPage):
+
+    NAME = "AcousticBrainz"
+    TITLE = N_("AcousticBrainz")
+    PARENT = None
+    SORT_ORDER = 45
+    ACTIVE = True
+    HELP_URL = '/config/options_acousticbrainz.html'
+
+    options = [
+        BoolOption("setting", "use_acousticbrainz", False),
+        TextOption("setting", "acousticbrainz_extractor", ""),
+    ]
+
+    def __init__(self, parent=None):
+        super().__init__(parent)
+        self._extractor_valid = True
+        self.ui = Ui_AcousticBrainzOptionsPage()
+        self.ui.setupUi(self)
+        self.ui.disable_acoustic_features.clicked.connect(self.update_groupboxes)
+        self.ui.use_acoustic_features.clicked.connect(self.update_groupboxes)
+        self.ui.acousticbrainz_extractor.textChanged.connect(self._acousticbrainz_extractor_check)
+        self.ui.acousticbrainz_extractor_browse.clicked.connect(self.acousticbrainz_extractor_browse)
+        self.ui.acousticbrainz_extractor_download.clicked.connect(self.acousticbrainz_extractor_download)
+        self._acousticbrainz_extractor_info = ""
+        self._tagger = QtCore.QCoreApplication.instance()
+        self._config = get_config()
+
+    def load(self):
+        if self._config.setting["use_acousticbrainz"]:
+            self.ui.use_acoustic_features.setChecked(True)
+        else:
+            self.ui.disable_acoustic_features.setChecked(True)
+        self.ui.acousticbrainz_extractor.setPlaceholderText("streaming_extractor_music")
+        self.ui.acousticbrainz_extractor.setText(self._config.setting["acousticbrainz_extractor"])
+        self.update_groupboxes()
+
+    def save(self):
+        enabled = self.ui.use_acoustic_features.isChecked()
+        self._config.setting["use_acousticbrainz"] = enabled
+        if enabled:
+            self._config.setting["acousticbrainz_extractor"] = self.ui.acousticbrainz_extractor.text()
+            ab_setup_extractor(self._tagger, self._config)
+
+    def update_groupboxes(self):
+        enabled = self.ui.use_acoustic_features.isChecked()
+        self.ui.acousticbrainz_settings.setEnabled(enabled)
+        self._acousticbrainz_extractor_check()
+
+    def acousticbrainz_extractor_browse(self):
+        path, _filter = QtWidgets.QFileDialog.getOpenFileName(self, "", self.ui.acousticbrainz_extractor.text())
+        if path:
+            path = os.path.normpath(path)
+            self.ui.acousticbrainz_extractor.setText(path)
+
+    def acousticbrainz_extractor_download(self):
+        webbrowser2.open(ACOUSTICBRAINZ_DOWNLOAD_URL)
+
+    def _acousticbrainz_extractor_check(self):
+        if not self.ui.use_acoustic_features.isChecked():
+            self._acousticbrainz_extractor_set_success("")
+            return
+
+        if not os.path.exists(self.ui.acousticbrainz_extractor.text()):
+            # If current path to extractor isn't valid, look for an extractor
+            self.ui.acousticbrainz_extractor.setText(find_extractor())
```

Comment:
This currently makes it impossible to manually change the path, as it immediately tries to correct it and guess it.

> I think I can just look for the configured path, and if it doesn't exist, just search again. Kind of defeats the purpose of setting the extractor, but whatever.

The issue is that for the portable version the path changes on each start. Similar issue happens with the Snap version, there the path changes with each app update.

For AcoustID we hence went with the solution that it defaults to an empty configured path, and when the path is empty it always gets guessed on invocation. See my changes in https://github.com/metabrainz/picard/pull/1724 . I would suggest to make the AB extractor detection work the same way. So instead of saving the auto detection to the config, allow an empty path and if it is empty auto detect when actually using the feature.

Reply:
> So instead of saving the auto detection to the config, allow an empty path and if it is empty auto detect when actually using the feature.

Yup, it works much better doing that. :D
I really didn't want to keep the field empty (which in is pretty of dumb, since there is a placeholder anyways...).

Reply:
We struggled with this for the fpcalc case as well. The original behavior that was implemented was that Picard would detect when it was running packaged by PyInstaller and if that was the case it auto detected on startup every time. But this had the downside that you could not intentionally configure an alternative path without loosing it. It also was unnecessary most of the time for macOS and Windows installer case and did not capture the snap case.

With "empty = autodetection" this actually now works well, and gives a clear option to have it auto detect (which probably covers most users needs anyway). I'd just make sure to not run  the auto detection too often. I think the fpcalc case only does auto detection once when you click scan.